### PR TITLE
feat(autocomplete): emit event when an option is selected

### LINKED ISSUE #3645: feat(autocomplete): fire "onSelected" event when the user select an option
#### Bug, feature request, or proposal:
featur

#### What is the expected behavior?
Create a new output (onSelected?) that fires an event when the user select an option from the autocomplete

#### What is the current behavior?
There is no way to know when the user has selected an option from the autocomplete

#### What is the use-case or motivation for changing an existing behavior?
I have some cases that I need to update some fields based on the user selected option and would be nice if I can get a callback from the component itself

### LINKED ISSUE #4094: Autocomplete onSelectionChange fires two times when changing previously selected value
#### Bug:
If the autocomplete input field already has a value and this is changed by deleting the previous value and selecting a new option from the list, then the autocomplete's `onSelectionChange` fires twice.

#### What is the expected behavior?
When editing/ changing the previous selected value `onSelectionChange` should fire only once.

#### What is the current behavior?
When editing/ changing the previous selected value `onSelectionChange` fires twice

#### What are the steps to reproduce?
Step 1: Select a value from autocomplete dropdown
Step 2: Clear this value (use backspace/ select all and erase)
Step 3: Select a different value from the autocomplete

Providing a Plunker (or similar) is the *best* way to get the team to see your issue.
https://plnkr.co/edit/cOYg48HGf86ZtG0fO59r?p=preview

#### What is the use-case or motivation for changing an existing behavior?

When you want to edit a pre-filled value or you selected a value by mistake and want to make a change and if there are other fields dependent on this value, then the dependent values don't get updated.

#### Which versions of Angular, Material, OS, browsers are affected?
"@angular/material": "^2.0.0-beta.3"
OS: Mac
browser: Chrome/ Safari

#### Is there anything else we should know?
Nope.

### PR DESCRIPTION
**Pull request by crisbeto**
Emits the `select` event when an option in the autocomplete is selected.

**Note:** I went with passing the selected option from the trigger to the panel, instead of listening to the `onSelectionChange` inside the panel, because it involves keeping track of less subscriptions and not having to re-construct them when the list of options changes.

Fixes #4094.
Fixes #3645.

### CODE REVIEW

File: src/lib/autocomplete/autocomplete.ts [hunk 1]
```diff
@@ -50,6 +58,9 @@ export class MdAutocomplete implements AfterContentInit {
   /** Function that maps an option's control value to its display value in the trigger. */
   @Input() displayWith: (value: any) => string;
 
+  /** Event that is emitted whenever an option from the list is selected. */
+  @Output() select: EventEmitter<MdAutocompleteSelect> = new EventEmitter<MdAutocompleteSelect>();
```
Comment (DevVersion) [inline src/lib/autocomplete/autocomplete.ts]: Shouldn't we just expose the `Observable` instead of the `EventEmitter`?

> Also as a side-question: What about a `change` event?

Comment (crisbeto) [inline src/lib/autocomplete/autocomplete.ts]: I wanted to avoid `change`, because typing in the select also changes the value. Regarding the event, using the `EventEmitter` is the way to handle `@Output`.

Comment (DevVersion) [inline src/lib/autocomplete/autocomplete.ts]: `@Output` can be basically any `Observable`. `EventEmitter` is just like a RxJS `Subscription`. We originally changed all `@Output`s to just expose the `Observable`. See https://github.com/angular/material2/issues/440

Comment (DevVersion) [inline src/lib/autocomplete/autocomplete.ts]: Actually it looks like a lot of components don't do that anymore. @jelbourn any reason why we didn't continue just exposing the observable?

Comment (crisbeto) [inline src/lib/autocomplete/autocomplete.ts]: This is pretty much the only use for the `EventEmitter`: https://angular.io/docs/ts/latest/api/core/index/EventEmitter-class.html

**Comment by rafaelss95** [comment]
What the status of this PR? Can it be merged please?

File: src/lib/autocomplete/autocomplete.ts (hunk 1 above)
Comment (willshowell) [inline src/lib/autocomplete/autocomplete.ts]: @crisbeto @DevVersion is this resolved? I don't want this PR to be forgotten!

Comment (kevincaradant) [inline src/lib/autocomplete/autocomplete.ts]: +1 to have fix this problem :). I would to know when this PR will be accepted. Thank you

File: src/lib/autocomplete/autocomplete.spec.ts [hunk 2]
```diff
@@ -1146,6 +1147,29 @@ describe('MdAutocomplete', () => {
         expect(panel.classList).toContain(visibleClass, `Expected panel to be visible.`);
       });
     }));
+
+  it('should call emit an event when an option is selected', fakeAsync(() => {
```
Comment (kara) [inline src/lib/autocomplete/autocomplete.spec.ts]: Typo: 'should call emit an event' -> 'should emit an event'

File: src/lib/autocomplete/autocomplete.ts [hunk 3]
```diff
@@ -9,16 +9,24 @@ import {
...
 
+
 /**
  * Autocomplete IDs need to be unique across components, so this counter exists outside of
  * the component definition.
  */
 let _uniqueAutocompleteIdCounter = 0;
 
+/** Event object that is emitted when an autocomplete option is selected */
+export class MdAutocompleteSelect {
```
Comment (kara) [inline src/lib/autocomplete/autocomplete.ts]: How about `MdAutocompleteSelectEvent` ? Without "Event", it almost sounds like a hybrid md-autocomplete / md-select directive.

File: src/lib/autocomplete/autocomplete.ts (hunk 1 above)
Comment (kara) [inline src/lib/autocomplete/autocomplete.ts]: @jelbourn and I discussed and we like `(optionSelected)`. It's a bit more explicit, and otherwise it sounds a bit too much like md-select.

File: src/lib/autocomplete/autocomplete.spec.ts [hunk 4]
```diff
@@ -1146,6 +1147,29 @@ describe('MdAutocomplete', () => {
...
+    options[1].click();
+    tick();
+    fixture.detectChanges();
+
+    expect(spy).toHaveBeenCalledTimes(1);
+
+    let event = spy.calls.mostRecent().args[0] as MdAutocompleteSelect;
+
+    expect(event.source).toBe(fixture.componentInstance.autocomplete);
+    expect(event.option.value).toBe('Washington');
```
Comment (kara) [inline src/lib/autocomplete/autocomplete.spec.ts]: Can you add a test for when the option list changes? e.g. emitting an event from a new option.

**Comment by crisbeto** [comment]
Thanks for the feedback @kara. It's addressed now, can you take another look?

File: src/lib/autocomplete/autocomplete.ts [hunk 5]
```diff
@@ -18,16 +18,26 @@ import {
...
  * the component definition.
  */
 let _uniqueAutocompleteIdCounter = 0;
 
+/** Event object that is emitted when an autocomplete option is selected */
+export class MdAutocompleteSelectedEvent {
+  constructor(public source: MdAutocomplete, public option: MdOption) { }
+}
+
+export type AutocompletePositionY = 'above' | 'below';
```
Comment (kara) [inline src/lib/autocomplete/autocomplete.ts]: Where is this being used?  Seems unrelated to this PR.
